proposals/0000-implicit-resource-binding.md

```diff
@@ -356,10 +356,129 @@ part of the HLSL language.
+
+The constructor will call Clang builtin function
+`__builtin_hlsl_create_handle_from_implicit_binding` and will pass along the
+`space`, `range` and `index` values, and it will also include the uninitialized
+resource handle. The type of the handle argument will be used to infer the
+specific resource handle type returned by the buildin function. This will happen
+in `SemaHLSL::CheckBuiltinFunctionCall` the same way we infer return types for
+HLSL intrinsic builtins based on the builtin arguments.
+
+```c++
+template <typename T> class RWBuffer<T> {
+  RWBuffer(unsigned space, int range, unsigned index) {
+    __handle = __builtin_hlsl_create_handle_from_implicit_binding(__handle, space, range, index);
+  }
+}
+```
+
+### Clang Codegen
+
+The builtin function `__builtin_hlsl_create_handle_from_implicit_binding` will be
+translated to LLVM intrinsic `llvm.{dx|spv}.resource.handlefromimplicitbinding`.
+
+The signature of the `@llvm.dx.resource.handlefromimplicitbinding` intrinsic will be:
+
+| Argument | Type |	Description | 
+|-|-|-|
+| return value | `target()` type|	A handle which can be operated on |
+| %order_id | i32 | Number uniquely identifying the unbound resource and declaration order | 
+| %space	|	i32 |	Virtual register space |
+| %range	|	i32 |	Range size of the binding |
+| %index	|	i32	| Index from the beginning of the binding |
+
+_Note: We might need to add a uniformity bit here, unless we can derive it from uniformity analysis._
+
+The `order_id` number will be generated by Clang Codegen in `GCHLSLRuntime`. It
+will be used to uniquely identify the unbound resource, and it will also reflect
+the order in which the resource has been declared. Implicit binding assignments
+depend on the order the resources were declared, so this will help the compiler
+to preserve the order. Additionally, it will also help the compiler distinquish
+between individual resources with the same type and binding range.
+
+### LLVM Binding Assignment Pass
+
+The implicit binding assignment for DirectX will happen in an LLVM pass
+`DXILResourceImplicitBindings`. The pass will run after all IR optimizations and
+will use the results of `DXILResourceBindingAnalysis`. It needs to run before
+any other pass that uses this analysis because it will be updating it with the
+newly assigned bindings.
+
+The pass will scan the module to gather all instances of
+`@llvm.dx.resource.handlefromimplicitbinding` calls and sort them by
+`%order_id`. Then it will process the list sequentially based on the
+`%order_id`, and for each group of calls operating on the same unique resource
+it will find an available resource binding based on the required resource class,
+range and space. It will then replace all of the
+`@llvm.dx.resource.handlefromimplicitbinding` calls and all of its uses with new
+`@llvm.dx.resource.handlefrombinding` calls and add the new binding to the
+`DXILResourceBindingAnalysis` map. 
```

Comment:
How will this effect Increment/Decrement counter? Specifically, the need to uniquely identify a resource from the handle. (https://github.com/llvm/llvm-project/issues/125126)

Right now we inspect the `handlefrombinding`  calls which will happen during `DXILResourceBindingAnalysis` (after the changes from https://github.com/llvm/wg-hlsl/blob/main/proposals/0022-resource-instance-analysis.md are implemented) which means it will see `handlefromimplicitbinding` instead.

Doesn't `handlefromimplicitbinding` have any meaningful impact on uniqueness?

Reply:
The `DXILResourceImplicitBindings` pass needs to run before any other passes that work with resources. There are two options how to handle the interaction with `DXILResourceBindingAnalysis`:
1. The pass will use the `DXILResourceBindingAnalysis` and update it with the newly assigned bindings
2. The pass will do its own collecting of existing `@llvm.dx.resource.handlefrombinding` calls just to create a "map" of what registers are occupied. The `DXILResourceBindingAnalysis` can run as requested by the follow-up passes after `DXILResourceImplicitBindings` pass is completed.

I think the correct way to do this is #2 because the `DXILResourceBindingAnalysis` is no longer going to be just about bindings but it will analyze the counter usage and other things. I will update the doc.

Reply:
sounds good thanks!